apps/system/lockscreen/js/component/lockscreen_basic_component.js

```diff
@@ -0,0 +1,162 @@
+    var nextState = new clazz(this);
+    var currentState = this._activeState;
+    this._activeState = nextState;
+    this.logger.transfer(currentState.configs.name, nextState.configs.name);
+    return currentState.stop()
+      .next(() => nextState.start());
+  };
+
+  /**
+   * Would receive resources from parent and *extends* the default one.
+   * After that, transfer to the next state, which is usually an initialization
+   * state, that would do lots of sync/async things to update the
+   * resources & properties.
+   *
+   * However, since basic component couldn't know what is the
+   * initialization state, so that the concrete component should
+   * implement the setup function, which would return the state after
+   * receive the component instance.
+   */
+  LockScreenBasicComponent.prototype.start = function(resources) {
+    this.logger.start(this.configs.logger);
+    if (resources) {
+      for (var key in this.resources) {
+        if ('undefined' !== resources[key]) {
+          this.resources[key] = resources[key];
+        }
+      }
+    }
+    // Get the initialization state and let it fetch & set all.
+    // 'initializeStateMachine', if Java doomed the world.
+    // (and this is ECMAScript, a language (partially) inspired by Scheme!).
+    this._activeState = this.setup();
+    return this._activeState.start();
+  };
+
+  /**
+   * Receive the compoennt instance and return a initialization state
+   * which would fetch resources and set properties for this component.
+   *
+   * The concrete component should implement this.
+   */
+  LockScreenBasicComponent.prototype.setup = function(component) {};
+
+  LockScreenBasicComponent.prototype.stop = function() {
+    return this._activeState.stop();
+  };
+
+  LockScreenBasicComponent.prototype.destroy = function() {
+    return this._activeState.destroy()
+      .next(() => { this.logger.stop(); });
+  };
+
+  LockScreenBasicComponent.prototype.live = function() {
+    return this._activeState.until('stop');
+  };
+
+  LockScreenBasicComponent.prototype.exist = function() {
+    return this._activeState.until('destroy');
+  };
+
+  /**
+   * Can command all sub-components with one method and its arguments.
+   * For example, to 'start', or 'stop' them.
+   * Will return a Promise only be resolved after all sub-components
+   * executed the command. For example:
+   *
+   * subcomponents: {
+   *    buttons: [ButtonFoo, ButtonBar]
+   *    submit: Submit
+   * }
+   * var promised = parent.waitComponent(parent.stop.bind(parent));
+   *
+   * The promised would be resolved only after ButtonFoo, ButtonBar and Submit
+   * are all stopped.
+   *
+   * And since for states the sub-components is delegated to Component,
+   * state should only command these sub-components via this method,
+   * or access them individually via the component instance set at the
+   * setup stage.
+   */
+  LockScreenBasicComponent.prototype.waitComponents = function(method, args) {
```

Comment:
It's necessary for our architecture because when parent component need to end itself, it must command children to finish themselves, too. You don't see the use case is because of the current one widget is only the ClockWidget, which is directly controlled by the legacy 'lockscreen.js', so there is no need to use this method yet.

I think the comments may be a bit confused, and I could fix it. And the use case now is in the `LockScreenConnectionInformation` widget, which actually has the children components and need to command them in this way. However, it would be a definitely high-risky step to land it within this patch.